### DISCUSSION TITLE
Public API for accessing connection pool details

### DISCUSSION THREAD

**Discussion by RobertCraigie** [encode/httpx#2966] [Ideas]
(Apologies if this already exists and I've just missed this)

There doesn't appear to be a public API for accessing any details about the number of open connections / requests in a connection pool.

I recently ran into a bug where we weren't properly releasing connections back to the connection pool in certain scenarios and I'd love to be able to write unit tests to cover this case and add debug logs without relying on an internal API.

For reference this is the internal API I used for testing:

```py
import httpx

client = httpx.Client()

# ...

transport = client._transport
assert isinstance(transport, httpx.HTTPTransport)

pool = transport._pool
print("Pool:", len(pool._pool))
print("Requests:", len(pool._requests))
```

This makes use of:
- `httpx.Client._transport`
- `httpx.HTTPTransport._pool`
- `httpcore.ConnectionPool._pool`
- `httpcore.ConnectionPool._requests`

The first private property access could be obviated by instantiating the transport directly but that unfortunately isn't possible in my case.

**Comment by rattrayalex** [comment]
Gentle bump, is this something you'd consider @tomchristie ?

**Reply by lovelydinosaur** [reply]
Heya, yup it's a great suggestion. I've also consider this in the past yep, and would suggest tackling this functionality step-by-step.

* Figure out what `__repr__` would be appropriate on `httpcore.ConnectionPool` in order to expose the state of the connection pool.
* Then expose the same on `httpx.HTTPTransport.__repr__`.
* Then ensure we're exposing the transport via a `client.transport` property.

I'd prefer this approach since I'd prefer we avoid exposing too much implementation detail directly as public API, while still making some of the information visible.

**Reply by rattrayalex** [reply]
Thanks Tom.

Interesting – when you say `__repr__`, do you mean to imply that data about the connection pool would only be accessible by string parsing?

Another use-case where we've run into this is that we may want to warn users when they're getting close to a connection pool limit in a live application. I wouldn't want production code to rely on parsing a connection count from a string.

Perhaps I misunderstand?

**Reply by lovelydinosaur** [reply]
Okay... Perhaps what I mean here is that I don't feel comfortable taking a decision that'd expose more API just yet, but that a `__repr__` exposing some information about the state of the pool would be a good proxy onto understand exactly what we would like to expose.

For example, is getting to the point where we can do this...

```python
>>> client = httpx.Client()
>>> client.transport
<httpx.HTTPTransport [Connections: 0 active, 0 keep-alive. Requests: 0 active, 0 queued.]>
```

...a decent/sufficient incremental starting point onto what you're looking for?

**Reply by rattrayalex** [reply]
Thank you Tom!

That would certainly be helpful in some cases. 

Exposing that information in a dictionary is what we would need in any kind of production use-case, such as altering program behavior or alerting when the number of available connections is getting dangerously low.

Without structured data, users in our shoes might be tempted to regex it out or just use the private variables, which wouldn't be great.

I haven't thought deeply about which properties exactly we'd need, but we'd take what we can get and the ones you show there look good at first blush.

Hope this is helpful!
Alex

**Reply by lovelydinosaur** [reply]
Lovely. I'd suggest that we work through this incrementally.

Perhaps start with an issue in `httpcore` describing adding `__repr__` to connection pools, and we can work forwards from there.

**Reply by rattrayalex** [reply]
Done: https://github.com/encode/httpcore/issues/861
